network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,390 @@
+    ) -> io::Result<NoiseSession<TSocket>>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // create prologue as current timestamp in seconds, and send it
+        let now: u64 = time::SystemTime::now()
+            .duration_since(time::UNIX_EPOCH)
+            .expect("system clock should work")
+            .as_secs();
+        let prologue = now.to_le_bytes(); // 5 -> [0, 0, 0, 0, 0, 0, 0, 5]
+        socket.write_all(&prologue).await?;
+
+        // create first handshake message  (-> e, es, s, ss)
+        let mut rng = rand::rngs::OsRng;
+        let mut first_message = [0u8; noise::handshake_init_msg_len(0)];
+        let initiator_state = self
+            .0
+            .initiate_connection(
+                &mut rng,
+                &prologue,
+                remote_public_key,
+                None,
+                &mut first_message,
+            )
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("noise: wrong input passed {}", e),
+                )
+            })?;
+
+        // write the first handshake message
+        socket.write_all(&first_message).await?;
+
+        // flush
+        socket.flush().await?;
+
+        // receive the server's response (<- e, ee, se)
+        let mut server_response = [0u8; noise::handshake_resp_msg_len(0)];
+        socket.read_exact(&mut server_response).await?;
+
+        // parse the server's response
+        // TODO: security logging here? (mimoo)
+        let (_, session) = self
+            .0
+            .finalize_connection(initiator_state, &server_response)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // finalize the connection
+        Ok(NoiseSession::new(socket, session))
+    }
+
+    pub async fn accept<TSocket>(
+        &self,
+        mut socket: TSocket,
+        anti_replay_timestamps: Arc<RwLock<AntiReplayTimestamp>>,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<NoiseSession<TSocket>>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // receives prologue as the client timestamp in seconds
+        let mut prologue = [0u8; 8];
+        socket.read_exact(&mut prologue).await?;
+        let client_timestamp_u64 = u64::from_le_bytes(prologue);
+        let client_timestamp = time::Duration::from_secs(client_timestamp_u64);
+
+        // check the client timestamp
+        // TODO: security logging (mimoo)
+        let now = time::SystemTime::now()
+            .duration_since(time::UNIX_EPOCH)
+            .expect("system clock should work");
+        if client_timestamp > now
+            && client_timestamp - now > time::Duration::from_secs(MAX_FUTURE_TIMESTAMP)
```

Comment:
what do you propose?

If we don't rely on the timestamp being correct then we do need to measure the drift to know when it can't be replayed OR we need to send them our timestamp before they start a handshake with us :D